File: skills/code-port/examples/go2ts/main.ts
// Port of main.go (concurrent URL fetcher) — Go -> TypeScript.
// Phase A draft, verified to type-check in Phase B (tsc --strict).
//
// Axis classification (S=Go, T=TypeScript):
//   Typing:      nominal interface -> structural interface (1:1 in TS).
//   Error:       (string, error) value-tuples -> throw + try/catch. fetchAll
//                still aggregates into Result.Err (value), preserving behavior;
//                only the *transport* (Fetch return) moved from value to throw.
//   Concurrency: goroutines + WaitGroup writing results[idx] -> Promise.all
//                (eager, single-threaded event loop). Concurrency preserved,
//                not parallelism (Node is single-threaded; Go was M:N). Flagged.
//   Cleanup:     `defer resp.Body.Close()` -> fetch drains the body automatically
//                (no manual close needed in the web fetch API).
//   Bytes:       Go len(string)=bytes; TS .length=UTF-16 units -> TODO(port).

interface Fetcher {
  fetch(url: string): Promise<string>;
}

class HttpFetcher implements Fetcher {
  async fetch(url: string): Promise<string> {
    // PORT NOTE: Go http.Get + io.ReadAll + defer Close; web fetch drains itself.
    const resp = await fetch(url);
    if (!resp.ok) {
      throw new Error(`HTTP ${resp.status}`);
    }
    return await resp.text();
  }
}

interface Result {
  url: string;
  body: string;
  err: Error | null;
}

async function fetchAll(f: Fetcher, urls: string[]): Promise<Result[]> {
  const results: Result[] = new Array<Result>(urls.length);
  // PORT NOTE: Go spawns a goroutine per url writing results[idx] under a
  // WaitGroup; here Promise.all awaits all in flight. Field order preserved.
  await Promise.all(
    urls.map(async (url, idx) => {
      try {
        const body = await f.fetch(url);
        results[idx] = { url, body, err: null };
      } catch (e) {
        results[idx] = {
          url,
          body: "",
          err: e instanceof Error ? e : new Error(String(e)),
        };
      }
    }),
  );
  return results;
}

async function main(): Promise<void> {
  const urls = ["https://example.com", "https://example.org"];
  const res = await fetchAll(new HttpFetcher(), urls);
  for (const r of res) {
    if (r.err) {
      console.log(`${r.url}: error: ${r.err.message}`);
      continue;
    }
    // TODO(port): Go len(string) counts bytes; TS .length counts UTF-16 code units.
    console.log(`${r.url}: ${r.body.length} chars`);
  }
}

main();

// ──────────────────────────────────────────────────────────────────────────
// PORT STATUS
//   source:     main.go (55 lines)
//   confidence: high
//   todos:      1 (byte-vs-char length on r.body)
//   notes:      error model mapped value->throw at the Fetch boundary, kept as
//               value (Result.err) at the aggregation boundary, matching Go.
//               Concurrency preserved; parallelism not (single-threaded runtime).
// ──────────────────────────────────────────────────────────────────────────

File: skills/code-port/examples/go2ts/main.go
package main

import (
	"fmt"
	"io"
	"net/http"
	"sync"
)

type Fetcher interface {
	Fetch(url string) (string, error)
}

type httpFetcher struct{}

func (httpFetcher) Fetch(url string) (string, error) {
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return string(body), nil
}

type Result struct {
	URL   string
	Body  string
	Err   error
}

func fetchAll(f Fetcher, urls []string) []Result {
	results := make([]Result, len(urls))
	var wg sync.WaitGroup
	for i, u := range urls {
		wg.Add(1)
		go func(idx int, url string) {
			defer wg.Done()
			body, err := f.Fetch(url)
			results[idx] = Result{URL: url, Body: body, Err: err}
		}(i, u)
	}
	wg.Wait()
	return results
}

func main() {
	urls := []string{"https://example.com", "https://example.org"}
	res := fetchAll(httpFetcher{}, urls)
	for _, r := range res {
		if r.Err != nil {
			fmt.Printf("%s: error: %v\n", r.URL, r.Err)
			continue
		}
		fmt.Printf("%s: %d bytes\n", r.URL, len(r.Body))
	}
}
